### PR TITLE
Document plugin directory structure

### PR DESCRIPTION
**Pull request by tsteur**
### Description:

Will later add a link from the https://developer.matomo.org/guides/how-piwik-works page to this new page. Thought I'd summarise a few files and directories. Hope I have the most important ones. This can be useful when people see a file but don't know what it is about. Then they find the related docs where they already exists. Also helps me to see where documentation is still missing.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: docs/4.x/plugin-directory-structure.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+|  tests/UI/  |     This directory includes [UI screenshot tests](/guides/tests-ui). Each test file ends with `_spec.js`.       |
+|  tests/Unit/  |     This directory includes unit [PHP tests](/guides/tests-php).       |
+|  tests/System/  |     This directory includes system [PHP tests](/guides/tests-php).       |
+|  tracker.js  |     If such a file is present, then this code will be added to the [Matomo JavaScript tracking code](/guides/enrich-js-tracker). If a `tracker.min.js` exists then the minified version will be used.   |
+|  Tracker/RequestProcessor.php  |     This file is typically used to [hook into various events during a tracking request](/guides/tracking-requests).      |
+|  Tracker/LogTable/  |     This directory can include log tables if the plugin creates custom log tables for tracking data.      |
+|  VisitorDetails.php  |    This file is used to add [additional information to the visitor log or visitor profile](/guides/visitor-log-and-profile).     |
+|  Visualizations/  |     This directory includes [visualizations](/guides/visualizing-report-data) that show report data differently.      |
```
Comment (sgiehl) [inline docs/4.x/plugin-directory-structure.md:57]: We could also mention `vendor/`, which is similar to `libs`, but holds composer requirements

File: docs/4.x/plugin-directory-structure.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+|  Tasks.php  |    Defines [scheduled tasks](/guides/scheduled-tasks) that are executed periodically similar to a cron.     |
+|  templates/ |     This directory includes twig template files.    |
+|  tests/Fixtures/  |     This directory includes all kind of fixtures for our automated tests.       |
+|  tests/Framework/Mock  |     This directory may include PHP mocks used in PHP tests.      |
+|  tests/Integration/  |     This directory includes integration [PHP tests](/guides/tests-php).       |
+|  tests/javascript/index.php  |     This file includes [JavaScript tests](/guides/jstracker-core#tests) for our JS tracker and Tag Manager.      |
+|  tests/javascript/head.php  |     This file can be used to load additional JS files in JavaScript tests for our JS tracker and Tag Manager.      |
+|  tests/UI/  |     This directory includes [UI screenshot tests](/guides/tests-ui). Each test file ends with `_spec.js`.       |
+|  tests/Unit/  |     This directory includes unit [PHP tests](/guides/tests-php).       |
+|  tests/System/  |     This directory includes system [PHP tests](/guides/tests-php).       |
```
Comment (sgiehl) [inline docs/4.x/plugin-directory-structure.md:51]: should we mention `tests/travis/*.yml`, which allows to append additional build steps to the generated .travis.yml?

File: docs/4.x/plugin-directory-structure.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+|  Archiver.php  |    This file includes logic on how to create archive reports from log raw data.     |
+|  Categories/  |     This directory includes menu categories and subcategories that are used to build the reporting menu. They define for example the order in which they are shown in the menu.    |
+|  Columns/ |     This directory includes [dimensions](/guides/dimensions) and metrics.    |
+|  Commands/ |     This directory includes [console commands](/guides/piwik-on-the-command-line).    |
+|  config/ |     This directory can include various files like "config.php" and "test.php" to change Matomo using [dependency injection](/guides/dependency-injection).    |
+|  Controller.php  |     Defines [controller actions](/guides/pages) for this plugin.    |
+|  Dao/  |     This directory includes data access objects that connect mostly to the database and execute queries. It doesn't contain much logic usually.    |
+|  DataTable/Filter  |     This directory may include [custom plugin data table filters](/guides/datatable#custom-filter).    |
+|  Diagnostic/  |     This directory typically includes [system checks](/guides/system-check).    |
```
Comment (sgiehl) [inline docs/4.x/plugin-directory-structure.md:24]: There might also be `LocationProvider/`, which can include additional geolocation providers.

File: docs/4.x/plugin-directory-structure.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+Each plugin has its own directory within the root `plugins/` directory. Generally, plugins have a similar [structure as the Matomo core](/guides/how-piwik-works) for most things. Most plugins have only few files, while some others have more files. Below is a list of some files and directories that may exist in a plugin.
+
+
+| File / Directory      | Description |
+| ----------- | ----------- |
+| plugin.json      | Contains [metadata about the plugin](/guides/distributing-your-plugin#prepare-your-plugin) such as contact information, a description, and more. Mostly needed for plugins on the Marketplace.      |
+|  $PluginName.php  |    `$PluginName` needs to be replaced with the name of the plugin. This file is the plugin file which allows plugins to listen to [events](/guides/events) and method hooks such as `install()`, `activate` and `uninstall()`.      |
+|  API.php  |    Defines [API methods](/guides/expose-api-methods) for this plugin.     |
+|  Archiver.php  |    This file includes logic on how to create archive reports from log raw data.     |
```
Comment (sgiehl) [inline docs/4.x/plugin-directory-structure.md:16]: should we mention `Activity/`, which holds possible activities for the ActivityLog Premium feature?

**Review by sgiehl — CHANGES_REQUESTED** [review]

File: docs/4.x/plugin-directory-structure.md (hunk 4 above)
Comment (tsteur) [reply docs/4.x/plugin-directory-structure.md:16]: done

**Pull request merged**